core/src/error.rs

```diff
@@ -38,11 +38,14 @@ pub enum Error {
 #[cfg(feature = "std")]
 impl std::error::Error for Error {}
 
-impl From<scale_decode::visitor::DecodeError> for Error {
-    fn from(value: scale_decode::visitor::DecodeError) -> Self {
-        Error::Decode(value.into())
-    }
-}
+convert_error!(ExtrinsicParamsError as Error, Error::ExtrinsicParams);
+convert_error!(BlockError as Error, Error::Block);
+convert_error!(MetadataError as Error, Error::Metadata);
+convert_error!(scale_decode::Error as Error, Error::Decode);
+convert_error!(scale_decode::visitor::DecodeError as Error, Error::Decode);
+convert_error!(scale_encode::Error as Error, Error::Encode);
+convert_error!(StorageAddressError as Error, Error::StorageAddress);
+convert_error!(codec::Error as Error, Error::Codec);
```

Comment:
Just a small thought to simplify this; one could do:

```rust
macro_rules! convert_error {
    ($module_path:path as $delegate_ty:ident :: $variant:ident) => {
        impl From<$module_path> for $delegate_ty {
            fn from(val: $module_path) -> Self {
                $delegate_ty::$variant(val.into())
            }
        }
    };
}

// usage:
convert_error!(FooError as Error::Foo);
```

Which looks a little more natural :)

Either way, I'd turn the `ty` into `path` because that's really what you're capturing here